projects/output-and-error-handling/main.go

```diff
@@ -0,0 +1,61 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+)
+
+func main() {
+	for {
+		resp, err := http.Get("http://localhost:8080")
+		if err != nil {
+			fmt.Println("Server not working at the moment!")
+			fmt.Fprint(os.Stderr, "Server not working at the moment! ", err, "\n")
+			return
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode == 429 {
+			// fmt.Printf("\n%#v\n", resp.Header)
+			fmt.Println("Please wait!")
+			retryAfterStr := resp.Header["Retry-After"][0]
+			// retryAfterNum, numErr := strconv.Atoi(retryAfterStr)
+			retryAfterNum, numErr := strconv.ParseFloat(retryAfterStr, 64)
+
+			var retryDuration time.Duration
```

Comment:
This is a great example of a place I'd extract a function. Being able to write something like:

```go
func parseRetryAfterHeader(header string) (time.Duration, error) {
    retryAfterNum, numErr := strconv.ParseFloat(retryAfterStr, 64)
    if numErr == nil {
        return time.Duration(float64(time.Second) * retryAfterNum), nil
    }
    // Attempt to parse as timestamp
}
```

reads much more clearly than having a series of nested `if`/`else`s for at least three reasons:
1. The "early return" means you only need to focus on one case at a time, rather than keeping track of all of the conversions that were attempted or errors that may have had
2. The goal of this code is to either return a `time.Duration` or an error - so having code which is framed in that way (returning a `time.Duration`), rather than code which is also in-line with the rest of the for loop, allows you to focus more on what it's doing (and also how the duration is used after!)
3. It's really easy to write tests for this function.